SoftwareGuide/Latex/Appendices/CodingStyleGuide.tex

```diff
@@ -1634,8 +1634,11 @@ \section{Initialization and Assignment}
 \code{\#\#nameOn}/\code{\#\#nameOff} method is recommended instead of directly
 manipulating the data member.
 
-Smart pointers need not to be initialized, since they initialize themselves
-to the \code{null} pointer, so they are the sole exception to the above rule.
+An exception is also made for low level utility classes for which data member initialization would
+cause a significant performance penalty. This is why for example \code{FixedArray::m_InternalArray}
+and \code{Index::m_InternalArray} do not have a default member initializer. But when a class has a
+static `New()` member function to instantiate objects (as defined by \code{itkNewMacro}), it should
+definitely follow the rule to have all its member variables initialized when they are declared.
```

Comment:
Thanks Jon, and sorry, maybe I wasn't clear at this point. My intention was to say here that, when a class has a static `New()` member function, the performance argument _against_ data member initialization does not apply. Because the extra performance cost of doing data member initialization during the creation of an object is assumed to be _neglectable_, compared to the cost of heap allocation that takes place during its `T::New()` call.

So I meant to clarify that the creation of an object by `T::New()` member function is relatively expensive, in that case, we certainly should not worry about the _possible_ extra performance cost of data initialization. On the other hand, it very much makes sense to exclude low level utility classes like `FixedArray<T, N>` and `Index<N>` from this guideline.

The consistency argument applies to treating data members of a raw pointer type and data members of a smart pointer type the same way, by adding in-class default member initializers to all of their declarations. I guess that's what you mean, right?

Reply:
> I guess that's what you mean, right?

Nope. I mean that, and as it was stated until now in the guide, smart pointers initialize themselves to the null pointer (regardless of whether their members are initialized or not, or regardless of how expensive the smart pointer initialization is). And that was used as an argument to not require to initialize them. Now, as mentioned in the related PR, consistency is also important, and makes things more readable as well if smart pointers are initialized with the class initializer braces. Maybe putting that in a separate paragraph reduced confusion if you think it adds confusion to your cost argument.